File: src/main/java/io/github/scaredsmods/scaredsfactions/common/network/packet/OpenScreenC2SPacket.java
package io.github.scaredsmods.scaredsfactions.common.network.packet;

import com.mojang.authlib.GameProfile;
import io.github.scaredsmods.scaredsfactions.api.common.faction.setting.AbstractFactionSetting;
import io.github.scaredsmods.scaredsfactions.client.screen.menu.*;
import io.github.scaredsmods.scaredsfactions.common.ScaredsFactionMod;
import io.github.scaredsmods.scaredsfactions.common.faction.Faction;
import io.github.scaredsmods.scaredsfactions.common.faction.FactionSavedData;
import io.github.scaredsmods.scaredsfactions.common.util.MessageUtil;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.ComponentSerialization;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.SimpleMenuProvider;
import net.neoforged.neoforge.network.handling.IPayloadContext;

import java.util.*;
import java.util.stream.Collectors;

public record OpenScreenC2SPacket(ModScreens screen, Component title) implements CustomPacketPayload {

    public static final CustomPacketPayload.Type<OpenScreenC2SPacket> TYPE = new Type<>(ScaredsFactionMod.id("open_screen_c2s"));

    public static final StreamCodec<RegistryFriendlyByteBuf, OpenScreenC2SPacket> STREAM_CODEC = StreamCodec.composite(
            ModScreens.STREAM_CODEC, OpenScreenC2SPacket::screen,
            ComponentSerialization.STREAM_CODEC, OpenScreenC2SPacket::title,
            OpenScreenC2SPacket::new
    );

    public static void handle(OpenScreenC2SPacket packet, IPayloadContext ctx) {
        ServerPlayer player = (ServerPlayer) ctx.player();
        FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
        Faction faction = data.getFactionFromPlayer(player.getUUID());
        if (faction == null) return;
        switch (packet.screen) {
            case CLOSE -> {
                player.closeContainer();
            }
            case MANAGE_FACTION -> {
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new ManageFactionMenu(containerId, playerInventory), packet.title));
            }
            case RENAME_FACTION -> {
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new RenameFactionMenu(containerId, playerInventory), packet.title));
            }
            case CONFIRM_TRANSFER -> {
                UUID targetUUID = faction.getPendingTransfer();
                if (targetUUID == null) return;
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new ConfirmTransferOwnershipMenu(containerId, playerInventory, targetUUID), packet.title), buf -> buf.writeUUID(targetUUID));
            }
            case CONFIRM_RESET_BEACON -> {
                if (!faction.hasBeacon()) {
                    player.sendSystemMessage(MessageUtil.Prefix.error("You must have a beacon to do this!"));
                    return;
                }
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new ConfirmResetBeaconPosMenu(containerId, playerInventory), packet.title));
            }
            case FACTION_SETTINGS -> player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new FactionSettingsMenu(containerId, playerInventory, faction.getSettings()), packet.title),
                    buf -> {
                            buf.writeVarInt(faction.getSettings().size());
                            for (AbstractFactionSetting<? , ?> setting : faction.getSettings()) {
                                buf.writeUtf(setting.getNbtId());
                                CompoundTag tag = new CompoundTag();
                                setting.save(tag);
                                buf.writeNbt(tag);
                            }
                    }
            );
            case TRANSFER_OWNERSHIP -> {
                Map<GameProfile, Faction.Rank> profileMembers = new HashMap<>();
                for (Map.Entry<UUID, Faction.Rank> entry : faction.getMembers().entrySet()) {
                    ServerPlayer onlineMember = Objects.requireNonNull(player.getServer()).getPlayerList().getPlayer(entry.getKey());
                    GameProfile profile = onlineMember != null ? onlineMember.getGameProfile()
                            : player.getServer().getProfileCache().get(entry.getKey()).orElse(new GameProfile(entry.getKey(), "Unknown"));
                    profileMembers.put(profile, entry.getValue());
                }

                Map<GameProfile, Faction.Rank> filteredMembers = profileMembers.entrySet().stream()
                        .filter(entry -> entry.getValue() == Faction.Rank.FIELD_MARSHAL)
                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new TransferOwnershipMenu(containerId, playerInventory, filteredMembers), packet.title), buf -> {
                    buf.writeInt(faction.getSettings().size());
                    for (Map.Entry<GameProfile, Faction.Rank> entry : filteredMembers.entrySet()) {
                        ByteBufCodecs.GAME_PROFILE.encode(buf, entry.getKey());
                        buf.writeEnum(entry.getValue());
                    }
                });
            }
            case VIEW_MEMBERS -> {
                Map<GameProfile, Faction.Rank> profileMembers = new LinkedHashMap<>();

                List<Map.Entry<UUID, Faction.Rank>> sortedMembers = faction.getMembers().entrySet().stream()
                        .sorted(Map.Entry.comparingByValue(Comparator.comparingInt(Faction.Rank::getId).reversed().thenComparing(Faction.Rank::getName)))
                        .toList();
                for (Map.Entry<UUID, Faction.Rank> entry : sortedMembers) {
                    ServerPlayer onlineMember = Objects.requireNonNull(player.getServer()).getPlayerList().getPlayer(entry.getKey());
                    GameProfile profile = onlineMember != null ? onlineMember.getGameProfile()
                            : player.getServer().getProfileCache().get(entry.getKey()).orElse(new GameProfile(entry.getKey(), "Unknown"));
                    profileMembers.put(profile, entry.getValue());
                }
                player.openMenu(new SimpleMenuProvider((containerId, playerInventory, player1) -> new TransferOwnershipMenu(containerId, playerInventory, profileMembers), packet.title), buf -> {
                    buf.writeInt(profileMembers.size());
                    for (Map.Entry<GameProfile, Faction.Rank> entry : profileMembers.entrySet()) {
                        ByteBufCodecs.GAME_PROFILE.encode(buf, entry.getKey());
                        buf.writeEnum(entry.getValue());
                    }
                });
            }
        }
    }

    @Override
    public Type<OpenScreenC2SPacket> type() {
        return TYPE;
    }
}

File: src/main/kotlin/io/github/scaredsmods/scaredsfactions/server/network/packet/OpenScreenC2SPacket.kt
/*
*  Copyright (C) 2026 ScaredRabbitNL
*
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU Lesser General Public License as published by
*  the Free Software Foundation, either version 3 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU Lesser General Public License for more details.
*
*  You should have received a copy of the GNU Lesser General Public License
*  along with this program. If not, see <https://www.gnu.org/licenses/>.
*/
package io.github.scaredsmods.scaredsfactions.server.network.packet

import com.mojang.authlib.GameProfile
import io.github.scaredsmods.scaredsfactions.api.server.network.packet.IAbstractFactionPacket
import io.github.scaredsmods.scaredsfactions.api.server.network.packet.PairPacket
import io.github.scaredsmods.scaredsfactions.client.screen.menu.*
import io.github.scaredsmods.scaredsfactions.common.faction.FactionSavedData
import io.github.scaredsmods.scaredsfactions.common.faction.Faction.Rank
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.world.SimpleMenuProvider
import net.minecraft.world.inventory.MenuConstructor
import net.minecraftforge.network.NetworkEvent
import net.minecraftforge.network.NetworkHooks
import java.util.function.Consumer
import java.util.function.Supplier
import kotlin.collections.HashMap
import kotlin.collections.MutableMap

class OpenScreenC2SPacket(private val screen : ModScreens, private val title: Component) : PairPacket<ModScreens, Component, OpenScreenC2SPacket>(
	screen,
	title,
	{ buf, modScreens -> buf.writeEnum(modScreens)},
	{ buf, component -> buf.writeComponent(component)}
	) {

	companion object : IAbstractFactionPacket.Decoder<OpenScreenC2SPacket> {
		override fun decode(buf: FriendlyByteBuf): OpenScreenC2SPacket {
			return OpenScreenC2SPacket(buf.readEnum(ModScreens::class.java), buf.readComponent())
		}
	}

	override fun handle(packet: OpenScreenC2SPacket, ctx: Supplier<NetworkEvent.Context>) {
		ctx.get().enqueueWork {
			val player = ctx.get().getSender() ?: return@enqueueWork
			val data = FactionSavedData.getSavedData(player.serverLevel())
			when (screen) {
				ModScreens.MANAGE_FACTION -> NetworkHooks.openScreen(
					player, SimpleMenuProvider(
						MenuConstructor { id, inv, _ -> ManageFactionMenu(id, inv) },
						title
					)
				)
				ModScreens.RENAME_FACTION -> NetworkHooks.openScreen(
					player, SimpleMenuProvider(
						MenuConstructor { id, inv, _ -> RenameFactionMenu(id, inv) },
						Component.literal("Rename Faction")
					)
				)
				ModScreens.TRANSFER_OWNERSHIP -> {
					val faction = data.getFactionFromPlayer(player.uuid) ?: return@enqueueWork
					val profileMembers: MutableMap<GameProfile, Rank> = HashMap()
					for (entry in faction.members.entries) {
						val onlineMember = player.server.playerList.getPlayer(entry.key)
						val profile = onlineMember?.gameProfile
							?: player.server.profileCache?.get(entry.key)?.orElse(GameProfile(entry.key, "Unknown"))
							?: GameProfile(entry.key, "Unknown")
						profileMembers[profile] = entry.value
					}
					val filteredMembers = profileMembers.filter { (_, rank) -> rank == Rank.FIELD_MARSHAL }
					NetworkHooks.openScreen(player, SimpleMenuProvider(
						MenuConstructor { id, inv, _ -> TransferOwnershipMenu(id, inv, filteredMembers) },
						Component.literal("Transfer Ownership")
					), Consumer { buf ->
						buf!!.writeInt(filteredMembers.size)
						for (entry in filteredMembers.entries) {
							buf.writeGameProfile(entry.key)
							buf.writeEnum(entry.value)
						}
					})
				}
				ModScreens.VIEW_MEMBERS -> {
					val faction = data.getFactionFromPlayer(player.uuid) ?: return@enqueueWork
					val profileMembers: MutableMap<GameProfile, Rank> = LinkedHashMap()
					val sortedMembers = faction.members.entries
						.sortedByDescending { it.value.id }
					for (entry in sortedMembers) {
						val onlineMember = player.server.playerList.getPlayer(entry.key)
						val profile = onlineMember?.gameProfile
							?: player.server.profileCache?.get(entry.key)?.orElse(GameProfile(entry.key, "Unknown"))
							?: GameProfile(entry.key, "Unknown")
						profileMembers[profile] = entry.value
					}
					NetworkHooks.openScreen(player, SimpleMenuProvider(
						MenuConstructor { id, inv, _ -> ViewMembersMenu(id, inv, profileMembers) },
						Component.literal("View Members")
					), Consumer { buf ->
						buf!!.writeInt(profileMembers.size)
						for (entry in profileMembers.entries) {
							buf.writeGameProfile(entry.key)
							buf.writeEnum(entry.value)
						}
					})
				}
				ModScreens.FACTION_SETTINGS -> {
					val faction = data.getFactionFromPlayer(player.uuid) ?: return@enqueueWork
					NetworkHooks.openScreen(player, SimpleMenuProvider(
						MenuConstructor { id, inv, _ -> FactionSettingsMenu(id, inv, faction.settings) },
						Component.literal("Faction Settings")
					), Consumer { buf ->
						buf!!.writeVarInt(faction.settings.size)
						for (setting in faction.settings) {
							buf.writeUtf(setting.nbtId)
							val tag = CompoundTag()
							setting.save(tag)
							buf.writeNbt(tag)
						}
					})
				}
				ModScreens.CONFIRM_TRANSFER -> {
					val faction = data.getFactionFromPlayer(player.uuid) ?: return@enqueueWork
					val targetUUID = faction.pendingTransfer ?: return@enqueueWork
					NetworkHooks.openScreen(player, SimpleMenuProvider(
						{ id, inv, _ -> ConfirmTransferOwnershipMenu(id, inv, targetUUID) },
						title
					)
					) { buf -> buf.writeUUID(targetUUID) }
				}
				ModScreens.CLOSE -> player.closeContainer()
			}
		}
		ctx.get().packetHandled = true
	}


}
